src/unix/udp.c

```diff
@@ -616,23 +639,19 @@ static int uv__setsockopt_maybe_char(uv_udp_t* handle,
   if (val < 0 || val > 255)
     return -EINVAL;
 
-  if (handle->flags & UV_HANDLE_IPV6)
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IPV6,
-                   option6,
-                   &arg,
-                   sizeof(arg));
-  else
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IP,
-                   option4,
-                   &arg,
-                   sizeof(arg));
-
-  if (r)
-    return -errno;
+/*
+ * On Solaris and derivatives such as SmartOS, the length of socket options
+ * is always sizeof(int), except for the IP_MULTICAST_LOOP and IP_MULTICAST_TTL
+ * options.
+ */
+#if defined(__sun)
+  if ((handle->flags & UV_HANDLE_IPV6) ||
```

Comment:
Here's a test quickly put together: https://gist.github.com/misterdjules/72b596e4462dbbdb2d0c that tests only values set via `uv__setsockopt_maybe_char`. The results are:
```
$ cc test-sockopt-opts-size.c -lsocket -lnsl
$ ./a.out 
socket option IP_MULTICAST_LOOP uses char sized values
socket option IP_TTL uses int sized values
socket option IP_MULTICAST_TTL uses char sized values
socket option IPV6_MULTICAST_LOOP uses int sized values
socket option IPV6_UNICAST_HOPS uses int sized values
socket option IPV6_MULTICAST_HOPS uses int sized values
$
```